File: example/rnn/lstm.py
# pylint:skip-file
import sys
sys.path.insert(0, "../../python")
import mxnet as mx
import numpy as np
from collections import namedtuple
import time
import math
LSTMState = namedtuple("LSTMState", ["c", "h"])
LSTMParam = namedtuple("LSTMParam", ["i2h_weight", "i2h_bias",
                                     "h2h_weight", "h2h_bias"])
LSTMModel = namedtuple("LSTMModel", ["rnn_exec", "symbol",
                                     "init_states", "last_states",
                                     "seq_data", "seq_labels", "seq_outputs",
                                     "param_blocks"])

def lstm(num_hidden, indata, prev_state, param, seqidx, layeridx, dropout=0.):
    """LSTM Cell symbol"""
    if dropout > 0.:
        indata = mx.sym.Dropout(data=indata, p=dropout)
    i2h = mx.sym.FullyConnected(data=indata,
                                weight=param.i2h_weight,
                                bias=param.i2h_bias,
                                num_hidden=num_hidden * 4,
                                name="t%d_l%d_i2h" % (seqidx, layeridx))
    h2h = mx.sym.FullyConnected(data=prev_state.h,
                                weight=param.h2h_weight,
                                bias=param.h2h_bias,
                                num_hidden=num_hidden * 4,
                                name="t%d_l%d_h2h" % (seqidx, layeridx))
    gates = i2h + h2h
    slice_gates = mx.sym.SliceChannel(gates, num_outputs=4,
                                      name="t%d_l%d_slice" % (seqidx, layeridx))
    in_gate = mx.sym.Activation(slice_gates[0], act_type="sigmoid")
    in_transform = mx.sym.Activation(slice_gates[1], act_type="tanh")
    forget_gate = mx.sym.Activation(slice_gates[2], act_type="sigmoid")
    out_gate = mx.sym.Activation(slice_gates[3], act_type="sigmoid")
    next_c = (forget_gate * prev_state.c) + (in_gate * in_transform)
    next_h = out_gate * mx.sym.Activation(next_c, act_type="tanh")
    return LSTMState(c=next_c, h=next_h)


def lstm_unroll(num_lstm_layer, seq_len, input_size,
                num_hidden, num_embed, num_label, dropout=0.):
    """unrolled lstm network"""
    # initialize the parameter symbols
    embed_weight=mx.sym.Variable("embed_weight")
    cls_weight = mx.sym.Variable("cls_weight")
    cls_bias = mx.sym.Variable("cls_bias")
    param_cells = []
    last_states = []
    for i in range(num_lstm_layer):
        param_cells.append(LSTMParam(i2h_weight = mx.sym.Variable("l%d_i2h_weight" % i),
                                      i2h_bias = mx.sym.Variable("l%d_i2h_bias" % i),
                                      h2h_weight = mx.sym.Variable("l%d_h2h_weight" % i),
                                      h2h_bias = mx.sym.Variable("l%d_h2h_bias" % i)))
        state = LSTMState(c=mx.sym.Variable("l%d_init_c" % i),
                          h=mx.sym.Variable("l%d_init_h" % i))
        last_states.append(state)
    assert(len(last_states) == num_lstm_layer)

    label = mx.sym.Variable("label")
    last_hidden = []
    for seqidx in range(seq_len):
        # embeding layer
        data = mx.sym.Variable("t%d_data" % seqidx)

        hidden = mx.sym.Embedding(data=data, weight=embed_weight,
                                  input_dim=input_size,
                                  output_dim=num_embed,
                                  name="t%d_embed" % seqidx)
        # stack LSTM
        for i in range(num_lstm_layer):
            if i==0:
                dp=0.
            else:
                dp = dropout
            next_state = lstm(num_hidden, indata=hidden,
                              prev_state=last_states[i],
                              param=param_cells[i],
                              seqidx=seqidx, layeridx=i, dropout=dp)
            hidden = next_state.h
            last_states[i] = next_state
        # decoder
        if dropout > 0.:
            hidden = mx.sym.Dropout(data=hidden, p=dropout)
        last_hidden.append(hidden)
    concat = mx.sym.Concat(*last_hidden, dim = 0)
    fc = mx.sym.FullyConnected(data=concat,
                               weight=cls_weight,
                               bias=cls_bias,
                               num_hidden=num_label)
    sm = mx.sym.SoftmaxOutput(data=fc, label=label, name="sm")
    out_prob = [sm]
    for i in range(num_lstm_layer):
        state = last_states[i]
        state = LSTMState(c=mx.sym.BlockGrad(state.c, name="l%d_last_c" % i),
                          h=mx.sym.BlockGrad(state.h, name="l%d_last_h" % i))
        last_states[i] = state

    unpack_c = [state.c for state in last_states]
    unpack_h = [state.h for state in last_states]
    list_all = out_prob + unpack_c + unpack_h
    return mx.sym.Group(list_all)


def is_param_name(name):
    return name.endswith("weight") or name.endswith("bias") or\
        name.endswith("gamma") or name.endswith("beta")


def setup_rnn_model(ctx,
                    num_lstm_layer, seq_len,
                    num_hidden, num_embed, num_label,
                    batch_size, input_size,
                    initializer, dropout=0.):
    """set up rnn model with lstm cells"""
    rnn_sym = lstm_unroll(num_lstm_layer=num_lstm_layer,
                          num_hidden=num_hidden,
                          seq_len=seq_len,
                          input_size=input_size,
                          num_embed=num_embed,
                          num_label=num_label,
                          dropout=dropout)
    arg_names = rnn_sym.list_arguments()

    input_shapes = {}
    for name in arg_names:
        if name.endswith("init_c") or name.endswith("init_h"):
            input_shapes[name] = (batch_size, num_hidden)
        elif name.endswith("data"):
            input_shapes[name] = (batch_size, )
        else:
            pass

    arg_shape, out_shape, aux_shape = rnn_sym.infer_shape(**input_shapes)
    arg_arrays = [mx.nd.zeros(s, ctx) for s in arg_shape]
    args_grad = {}
    for shape, name in zip(arg_shape, arg_names):
        if is_param_name(name):
            args_grad[name] = mx.nd.zeros(shape, ctx)

    rnn_exec = rnn_sym.bind(ctx=ctx, args=arg_arrays,
                            args_grad=args_grad,
                            grad_req="add")
    param_blocks = []
    arg_dict = dict(zip(arg_names, rnn_exec.arg_arrays))
    for i, name in enumerate(arg_names):
        if is_param_name(name):
            initializer(name, arg_dict[name])

            param_blocks.append((i, arg_dict[name], args_grad[name], name))
        else:
            assert name not in args_grad
    out_dict = dict(zip(rnn_sym.list_outputs(), rnn_exec.outputs))

    init_states = [LSTMState(c=arg_dict["l%d_init_c" % i],
                             h=arg_dict["l%d_init_h" % i]) for i in range(num_lstm_layer)]
    seq_labels = rnn_exec.arg_dict["label"]
    seq_data = [rnn_exec.arg_dict["t%d_data" % i] for i in range(seq_len)]
    last_states = [LSTMState(c=out_dict["l%d_last_c_output" % i],
                             h=out_dict["l%d_last_h_output" % i]) for i in range(num_lstm_layer)]
    seq_outputs = out_dict["sm_output"]

    return LSTMModel(rnn_exec=rnn_exec, symbol=rnn_sym,
                     init_states=init_states, last_states=last_states,
                     seq_data=seq_data, seq_labels=seq_labels, seq_outputs=seq_outputs,
                     param_blocks=param_blocks)



def set_rnn_inputs(m, X, begin):
    seq_len = len(m.seq_data)
    batch_size = m.seq_data[0].shape[0]
    for seqidx in range(seq_len):
        idx = (begin + seqidx) % X.shape[0]
        next_idx = (begin + seqidx + 1) % X.shape[0]
        x = X[idx, :]
        y = X[next_idx, :]
        mx.nd.array(x).copyto(m.seq_data[seqidx])
        m.seq_labels[seqidx*batch_size : seqidx*batch_size+batch_size] = y

def calc_nll(seq_label_probs, X, begin):
    nll = -np.sum(np.log(seq_label_probs.asnumpy())) / len(X[0,:])
    return nll

def train_lstm(model, X_train_batch, X_val_batch,
               num_round, update_period,
               optimizer='rmsprop', half_life=2,max_grad_norm = 5.0, **kwargs):
    print("Training with train.shape=%s" % str(X_train_batch.shape))
    print("Training with val.shape=%s" % str(X_val_batch.shape))
    m = model
    seq_len = len(m.seq_data)
    batch_size = m.seq_data[0].shape[0]
    print("batch_size=%d" % batch_size)
    print("seq_len=%d" % seq_len)

    opt = mx.optimizer.create(optimizer,
                              **kwargs)

    updater = mx.optimizer.get_updater(opt)
    epoch_counter = 0
    log_period = max(1000 / seq_len, 1)
    last_perp = 10000000.0
    for iteration in range(num_round):
        nbatch = 0
        train_nll = 0
        # reset states
        for state in m.init_states:
            state.c[:] = 0.0
            state.h[:] = 0.0
        tic = time.time()
        assert X_train_batch.shape[0] % seq_len == 0
        assert X_val_batch.shape[0] % seq_len == 0
        for begin in range(0, X_train_batch.shape[0], seq_len):
            set_rnn_inputs(m, X_train_batch, begin=begin)
            m.rnn_exec.forward(is_train=True)
            # probability of each label class, used to evaluate nll
            seq_label_probs = mx.nd.choose_element_0index(m.seq_outputs,m.seq_labels)
            m.rnn_exec.backward()
            # transfer the states
            for init, last in zip(m.init_states, m.last_states):
                last.c.copyto(init.c)
                last.h.copyto(init.h)
            # update epoch counter
            epoch_counter += 1
            if epoch_counter % update_period == 0:
                # updare parameters
                norm = 0.
                for idx, weight, grad, name in m.param_blocks:
                    grad /= batch_size
                    l2_norm = mx.nd.norm(grad).asscalar()
                    norm += l2_norm*l2_norm
                norm = math.sqrt(norm)
                for idx, weight, grad, name in m.param_blocks:
                    if norm > max_grad_norm:
                        grad *= (max_grad_norm / norm)
                    updater(idx, grad, weight)
                    # reset gradient to zero
                    grad[:] = 0.0
            train_nll += calc_nll(seq_label_probs, X_train_batch, begin=begin)

            nbatch = begin + seq_len
            if epoch_counter % log_period == 0:
                print("Epoch [%d] Train: NLL=%.3f, Perp=%.3f" % (
                    epoch_counter, train_nll / nbatch, np.exp(train_nll / nbatch)))
        # end of training loop
        toc = time.time()
        print("Iter [%d] Train: Time: %.3f sec, NLL=%.3f, Perp=%.3f" % (
            iteration, toc - tic, train_nll / nbatch, np.exp(train_nll / nbatch)))

        val_nll = 0.0
        # validation set, reset states
        for state in m.init_states:
            state.c[:] = 0.0
            state.h[:] = 0.0
        for begin in range(0, X_val_batch.shape[0], seq_len):
            set_rnn_inputs(m, X_val_batch, begin=begin)
            m.rnn_exec.forward(is_train=False)
            # probability of each label class, used to evaluate nll
            seq_label_probs = mx.nd.choose_element_0index(m.seq_outputs,m.seq_labels)
            # transfer the states
            for init, last in zip(m.init_states, m.last_states):
                last.c.copyto(init.c)
                last.h.copyto(init.h)
            val_nll += calc_nll(seq_label_probs, X_val_batch, begin=begin)
        nbatch = X_val_batch.shape[0]
        perp = np.exp(val_nll / nbatch)
        print("Iter [%d] Val: NLL=%.3f, Perp=%.3f" % (
            iteration, val_nll / nbatch, np.exp(val_nll / nbatch)))
        if last_perp - 1.0 < perp:
            opt.lr *= 0.5
            print("Reset learning rate to %g" % opt.lr)
        last_perp = perp

def setup_rnn_sample_model(ctx,
                           params,
                           num_lstm_layer,
                           num_hidden, num_embed, num_label,
                           batch_size, input_size):
    seq_len = 1
    rnn_sym = lstm_unroll(num_lstm_layer=num_lstm_layer,
                          input_size=input_size,
                          num_hidden=num_hidden,
                          seq_len=seq_len,
                          num_embed=num_embed,
                          num_label=num_label)
    arg_names = rnn_sym.list_arguments()
    input_shapes = {}
    for name in arg_names:
        if name.endswith("init_c") or name.endswith("init_h"):
            input_shapes[name] = (batch_size, num_hidden)
        elif name.endswith("data"):
            input_shapes[name] = (batch_size, )
        else:
            pass
    arg_shape, out_shape, aux_shape = rnn_sym.infer_shape(**input_shapes)
    arg_arrays = [mx.nd.zeros(s, ctx) for s in arg_shape]
    arg_dict = dict(zip(arg_names, arg_arrays))
    for name, arr in params.items():
        arg_dict[name][:] = arr
    rnn_exec = rnn_sym.bind(ctx=ctx, args=arg_arrays, args_grad=None, grad_req="null")
    out_dict = dict(zip(rnn_sym.list_outputs(), rnn_exec.outputs))
    param_blocks = []
    params_array = list(params.items())
    for i in range(len(params)):
        param_blocks.append((i, params_array[i][1], None, params_array[i][0]))
    init_states = [LSTMState(c=arg_dict["l%d_init_c" % i],
                             h=arg_dict["l%d_init_h" % i]) for i in range(num_lstm_layer)]
    seq_labels = rnn_exec.arg_dict["label"]
    seq_data = [rnn_exec.arg_dict["t%d_data" % i] for i in range(seq_len)]
    last_states = [LSTMState(c=out_dict["l%d_last_c_output" % i],
                             h=out_dict["l%d_last_h_output" % i]) for i in range(num_lstm_layer)]
    seq_outputs = out_dict["sm_output"]

    return LSTMModel(rnn_exec=rnn_exec, symbol=rnn_sym,
                     init_states=init_states, last_states=last_states,
                     seq_data=seq_data, seq_labels=seq_labels, seq_outputs=seq_outputs,
                     param_blocks=param_blocks)

# Python3 np.random.choice is too strict in eval float probability so we use an alternative
import random
import bisect
import collections

def _cdf(weights):
    total = sum(weights)
    result = []
    cumsum = 0
    for w in weights:
        cumsum += w
        result.append(cumsum / total)
    return result

def _choice(population, weights):
    assert len(population) == len(weights)
    cdf_vals = _cdf(weights)
    x = random.random()
    idx = bisect.bisect(cdf_vals, x)
    return population[idx]

def sample_lstm(model, X_input_batch, seq_len, temperature=1., sample=True):
    m = model
    vocab = m.seq_outputs.shape[1]
    batch_size = m.seq_data[0].shape[0]
    outputs_ndarray = mx.nd.zeros(m.seq_outputs.shape)
    outputs_batch = []
    tmp = [i for i in range(vocab)]
    for i in range(seq_len):
        outputs_batch.append(np.zeros(X_input_batch.shape))
    for i in range(seq_len):
        set_rnn_inputs(m, X_input_batch, 0)
        m.rnn_exec.forward(is_train=False)
        outputs_ndarray[:] = m.seq_outputs
        for init, last in zip(m.init_states, m.last_states):
            last.c.copyto(init.c)
            last.h.copyto(init.h)
        prob = np.clip(outputs_ndarray.asnumpy(), 1e-6, 1 - 1e-6)
        if sample:
            rescale = np.exp(np.log(prob) / temperature)
            for j in range(batch_size):
                p = rescale[j, :]
                p[:] /= p.sum()
                outputs_batch[i][j] = _choice(tmp, p)
                # outputs_batch[i][j] = np.random.choice(vocab, 1, p)
        else:
            outputs_batch[i][:] = np.argmax(prob, axis=1)
        X_input_batch[:] = outputs_batch[i]
    return outputs_batch

File: example/rnn/lstm.R
require(mxnet)

# lstm cell symbol
lstm <- function(num.hidden, indata, prev.state, param, seqidx, layeridx, dropout=0) {
    if (dropout > 0)
        indata <- mx.symbol.Dropout(data=indata, p=dropout)
    i2h <- mx.symbol.FullyConnected(data=indata,
                                    weight=param$i2h.weight,
                                    bias=param$i2h.bias,
                                    num.hidden=num.hidden * 4,
                                    name=paste0("t", seqidx, ".l", layeridx, ".i2h"))
    h2h <- mx.symbol.FullyConnected(data=prev.state$h,
                                    weight=param$h2h.weight,
                                    bias=param$h2h.bias,
                                    num.hidden=num.hidden * 4,
                                    name=paste0("t", seqidx, ".l", layeridx, ".h2h"))
    gates <- i2h + h2h
    slice.gates <- mx.symbol.SliceChannel(gates, num.outputs=4,
                                          name=paste0("t", seqidx, ".l", layeridx, ".slice"))

    in.gate <- mx.symbol.Activation(slice.gates[[1]], act.type="sigmoid")
    in.transform <- mx.symbol.Activation(slice.gates[[2]], act.type="tanh")
    forget.gate <- mx.symbol.Activation(slice.gates[[3]], act.type="sigmoid")
    out.gate <- mx.symbol.Activation(slice.gates[[4]], act.type="sigmoid")
    next.c <- (forget.gate * prev.state$c) + (in.gate * in.transform)
    next.h <- out.gate * mx.symbol.Activation(next.c, act.type="tanh")
    
    return (list(c=next.c, h=next.h))
}

# unrolled lstm network
lstm.unroll <- function(num.lstm.layer, seq.len, input.size,
                        num.hidden, num.embed, num.label, dropout=0.) {

    embed.weight <- mx.symbol.Variable("embed.weight")
    cls.weight <- mx.symbol.Variable("cls.weight")
    cls.bias <- mx.symbol.Variable("cls.bias")
    param.cells <- list()
    last.states <- list()
    for (i in 1:num.lstm.layer) {
        param.cells[[i]] <- list(i2h.weight = mx.symbol.Variable(paste0("l", i, ".i2h.weight")),
                                 i2h.bias = mx.symbol.Variable(paste0("l", i, ".i2h.bias")),
                                 h2h.weight = mx.symbol.Variable(paste0("l", i, ".h2h.weight")),
                                 h2h.bias = mx.symbol.Variable(paste0("l", i, ".h2h.bias")))
        state <- list(c=mx.symbol.Variable(paste0("l", i, ".init.c")),
                      h=mx.symbol.Variable(paste0("l", i, ".init.h")))
        last.states[[i]] <- state
    }

    last.hidden <- list()
    label <- mx.symbol.Variable("label")
    for (seqidx in 1:seq.len) {
        # embeding layer
        data <- mx.symbol.Variable(paste0("t", seqidx, ".data"))

        hidden <- mx.symbol.Embedding(data=data, weight=embed.weight,
                                      input.dim=input.size,
                                      output.dim=num.embed,
                                      name=paste0("t", seqidx, ".embed"))
        
        # stack lstm
        for (i in 1:num.lstm.layer) {
            if (i==0) {
                dp <- 0
            }
            else {
                dp <- dropout
            }
            next.state <- lstm(num.hidden, indata=hidden,
                               prev.state=last.states[[i]],
                               param=param.cells[[i]],
                               seqidx=seqidx, layeridx=i, 
                               dropout=dp)
            hidden <- next.state$h
            last.states[[i]] <- next.state
        }
        # decoder
        if (dropout > 0)  
            hidden <- mx.symbol.Dropout(data=hidden, p=dropout)
        last.hidden <- c(last.hidden, hidden)
    }
    last.hidden$dim <- 0
    last.hidden$num.args <- seq.len
    concat <-mxnet:::mx.varg.symbol.Concat(last.hidden)
    fc <- mx.symbol.FullyConnected(data=concat,
                                   weight=cls.weight,
                                   bias=cls.bias,
                                   num.hidden=num.label)
    loss.all <- mx.symbol.SoftmaxOutput(data=fc, label=label, name="sm")
    unpack.c <- list()
    unpack.h <- list()
    for (i in 1:num.lstm.layer) {
         state <- last.states[[i]]
         state <- list(c=mx.symbol.BlockGrad(state$c, name=paste0("l", i, ".last.c")),
                            h=mx.symbol.BlockGrad(state$h, name=paste0("l", i, ".last.h" )))
        last.states[[i]] <- state
        unpack.c <- c(unpack.c, state$c)
        unpack.h <- c(unpack.h, state$h)
    }
    list.all <- c(loss.all, unpack.c, unpack.h)

    return (mx.symbol.Group(list.all))
}

is.param.name <- function(name) {
    return (grepl('weight$', name) || grepl('bias$', name) || 
           grepl('gamma$', name) || grepl('beta$', name) )
}

mx.model.init.params <- function(symbol, input.shape, initializer, ctx) {
  if (!is.mx.symbol(symbol)) stop("symbol need to be MXSymbol")
  slist <- symbol$infer.shape(input.shape)
  if (is.null(slist)) stop("Not enough information to get shapes")
  arg.params <- mx.init.create(initializer, slist$arg.shapes, ctx, skip.unknown=TRUE)
  aux.params <- mx.init.create(initializer, slist$aux.shapes, ctx, skip.unknown=FALSE)
  return(list(arg.params=arg.params, aux.params=aux.params))
}

# set up rnn model with lstm cells
setup.rnn.model <- function(ctx,
                            num.lstm.layer, seq.len,
                            num.hidden, num.embed, num.label,
                            batch.size, input.size,
                            initializer=mx.init.uniform(0.01), 
                            dropout=0) {

    rnn.sym <- lstm.unroll(num.lstm.layer=num.lstm.layer,
                           num.hidden=num.hidden,
                           seq.len=seq.len,
                           input.size=input.size,
                           num.embed=num.embed,
                           num.label=num.label,
                           dropout=dropout)
    arg.names <- rnn.sym$arguments
    input.shapes <- list()
    for (name in arg.names) {
        if (grepl('init.c$', name) || grepl('init.h$', name)) {
            input.shapes[[name]] <- c(num.hidden, batch.size)
        }
        else if (grepl('data$', name)) {
            input.shapes[[name]] <- c(batch.size)
        }
    }
    
    params <- mx.model.init.params(rnn.sym, input.shapes, initializer, ctx)

    args <- input.shapes
    args$symbol <- rnn.sym
    args$ctx <- ctx
    args$grad.req <- "add"
    rnn.exec <- do.call(mx.simple.bind, args)

    mx.exec.update.arg.arrays(rnn.exec, params$arg.params, match.name=TRUE)
    mx.exec.update.aux.arrays(rnn.exec, params$aux.params, match.name=TRUE)

    grad.arrays <- list()
    for (name in names(rnn.exec$ref.grad.arrays)) {
        if (is.param.name(name))
            grad.arrays[[name]] <- rnn.exec$ref.arg.arrays[[name]]*0
    }
    mx.exec.update.grad.arrays(rnn.exec, grad.arrays, match.name=TRUE)

    return (list(rnn.exec=rnn.exec, symbol=rnn.sym,
                 num.lstm.layer=num.lstm.layer, num.hidden=num.hidden, 
                 seq.len=seq.len, batch.size=batch.size,
                 num.embed=num.embed))

}


get.rnn.inputs <- function(m, X, begin) {
    seq.len <- m$seq.len
    batch.size <- m$batch.size
    seq.labels <- array(0, dim=c(seq.len*batch.size))
    seq.data <- list()
    for (seqidx in 1:seq.len) {
        idx <- (begin + seqidx - 1) %% dim(X)[2] + 1
        next.idx <- (begin + seqidx) %% dim(X)[2] + 1
        x <- X[, idx]
        y <- X[, next.idx]

        seq.data[[paste0("t", seqidx, ".data")]] <- mx.nd.array(as.array(x))
        seq.labels[((seqidx-1)*batch.size+1) : (seqidx*batch.size)] <- y
    }
    seq.data$label <- mx.nd.array(seq.labels)
    return (seq.data)
}


calc.nll <- function(seq.label.probs, X, begin) {
    nll = - sum(log(seq.label.probs)) / length(X[,1])
    return (nll)
}

train.lstm <- function(model, X.train.batch, X.val.batch,
                       num.round, update.period,
                       optimizer='sgd', half.life=2, max.grad.norm = 5.0, ...) {
    X.train.batch.shape <- dim(X.train.batch)
    X.val.batch.shape <- dim(X.val.batch)
    cat(paste0("Training with train.shape=(", paste0(X.train.batch.shape, collapse=","), ")"), "\n")
    cat(paste0("Training with val.shape=(", paste0(X.val.batch.shape, collapse=","), ")"), "\n")

    m <- model
    seq.len <- m$seq.len
    batch.size <- m$batch.size
    num.lstm.layer <- m$num.lstm.layer
    num.hidden <- m$num.hidden

    opt <- mx.opt.create(optimizer, rescale.grad=(1/batch.size), ...)

    updater <- mx.opt.get.updater(opt, m$rnn.exec$ref.arg.arrays)
    epoch.counter <- 0
    log.period <- max(as.integer(1000 / seq.len), 1)
    last.perp <- 10000000.0

    for (iteration in 1:num.round) {
        nbatch <- 0
        train.nll <- 0
        # reset states
        init.states <- list()
        for (i in 1:num.lstm.layer) {
            init.states[[paste0("l", i, ".init.c")]] <- mx.nd.zeros(c(num.hidden, batch.size))
            init.states[[paste0("l", i, ".init.h")]] <- mx.nd.zeros(c(num.hidden, batch.size))
        }
        mx.exec.update.arg.arrays(m$rnn.exec, init.states, match.name=TRUE) 

        tic <- Sys.time()

        stopifnot(dim(X.train.batch)[[2]] %% seq.len == 0)
        stopifnot(dim(X.val.batch)[[2]] %% seq.len == 0)

        for (begin in seq(1, dim(X.train.batch)[2], seq.len)) {
            # set rnn input
            rnn.input <- get.rnn.inputs(m, X.train.batch, begin=begin)
            mx.exec.update.arg.arrays(m$rnn.exec, rnn.input, match.name=TRUE) 

            mx.exec.forward(m$rnn.exec, is.train=TRUE)
            # probability of each label class, used to evaluate nll
            seq.label.probs <- mx.nd.choose.element.0index(m$rnn.exec$outputs[["sm_output"]], m$rnn.exec$arg.arrays[["label"]])
            mx.exec.backward(m$rnn.exec)
            # transfer the states
            init.states <- list()
            for (i in 1:num.lstm.layer) {
                init.states[[paste0("l", i, ".init.c")]] <- m$rnn.exec$outputs[[paste0("l", i, ".last.c_output")]]
                init.states[[paste0("l", i, ".init.h")]] <- m$rnn.exec$outputs[[paste0("l", i, ".last.h_output")]]
            }
            mx.exec.update.arg.arrays(m$rnn.exec, init.states, match.name=TRUE) 
            # update epoch counter
            epoch.counter <- epoch.counter + 1
            if (epoch.counter %% update.period == 0) {

                # the gradient of initial c and inital h should be zero
                init.grad <- list()
                for (i in 1:num.lstm.layer) {
                    init.grad[[paste0("l", i, ".init.c")]] <- m$rnn.exec$ref.arg.arrays[[paste0("l", i, ".init.c")]]*0
                    init.grad[[paste0("l", i, ".init.h")]] <- m$rnn.exec$ref.arg.arrays[[paste0("l", i, ".init.h")]]*0
                }
                mx.exec.update.grad.arrays(m$rnn.exec, init.grad, match.name=TRUE)

                arg.blocks <- updater(m$rnn.exec$ref.arg.arrays, m$rnn.exec$ref.grad.arrays)

                mx.exec.update.arg.arrays(m$rnn.exec, arg.blocks, skip.null=TRUE)

                grad.arrays <- list()
                for (name in names(m$rnn.exec$ref.grad.arrays)) {
                    if (is.param.name(name))
                        grad.arrays[[name]] <- m$rnn.exec$ref.grad.arrays[[name]]*0
                }
                mx.exec.update.grad.arrays(m$rnn.exec, grad.arrays, match.name=TRUE)

            }

            train.nll <- train.nll + calc.nll(as.array(seq.label.probs), X.train.batch, begin=begin)

            nbatch <- begin + seq.len
            if ((epoch.counter %% log.period) == 0) {
                cat(paste0("Epoch [", epoch.counter, 
                           "] Train: NLL=", train.nll / nbatch, 
                           ", Perp=", exp(train.nll / nbatch), "\n"))
            }
        }
        # end of training loop
        toc <- Sys.time()
        cat(paste0("Iter [", iteration, 
                   "] Train: Time: ", as.numeric(toc - tic, units="secs"),
                   " sec, NLL=", train.nll / nbatch,
                   ", Perp=", exp(train.nll / nbatch), "\n"))

        val.nll <- 0.0
        # validation set, reset states
        init.states <- list()
        for (i in 1:num.lstm.layer) {
            init.states[[paste0("l", i, ".init.c")]] <- mx.nd.zeros(c(num.hidden, batch.size))
            init.states[[paste0("l", i, ".init.h")]] <- mx.nd.zeros(c(num.hidden, batch.size))
        }
        mx.exec.update.arg.arrays(m$rnn.exec, init.states, match.name=TRUE) 

        for (begin in seq(1, dim(X.val.batch)[2], seq.len)) {
            # set rnn input
            rnn.input <- get.rnn.inputs(m, X.val.batch, begin=begin)
            mx.exec.update.arg.arrays(m$rnn.exec, rnn.input, match.name=TRUE) 
            mx.exec.forward(m$rnn.exec, is.train=FALSE)
            # probability of each label class, used to evaluate nll
            seq.label.probs <- mx.nd.choose.element.0index(m$rnn.exec$outputs[["sm_output"]], m$rnn.exec$arg.arrays[["label"]])
            # transfer the states
            init.states <- list()
            for (i in 1:num.lstm.layer) {
                init.states[[paste0("l", i, ".init.c")]] <- m$rnn.exec$outputs[[paste0("l", i, ".last.c_output")]]
                init.states[[paste0("l", i, ".init.h")]] <- m$rnn.exec$outputs[[paste0("l", i, ".last.h_output")]]
            }
            mx.exec.update.arg.arrays(m$rnn.exec, init.states, match.name=TRUE)
            val.nll <- val.nll + calc.nll(as.array(seq.label.probs), X.val.batch, begin=begin)
        }
        nbatch <- dim(X.val.batch)[2]
        perp <- exp(val.nll / nbatch)
        cat(paste0("Iter [", iteration,  
                   "] Val: NLL=", val.nll / nbatch,
                   ", Perp=", exp(val.nll / nbatch), "\n"))


    }
}